### ISSUE TITLE
Creating a plugin for SDMX data sources

### ISSUE THREAD

**Issue by dr-leo** [intake/intake#593]
Hi,

I am the original author and current maintainer of (pandaSDMX)[http://pandasdmx.readthedocs.io], a library to explore, acquire and generate SDMX data. Supported sources are institutions such as central banks, national stats offices and international organizations.

I like the ecosystem around intake, panel and friends and want to see SDMX data therein.

I have just published the first commits on intake_sdmx at

https://github.com/dr-leo/intake_sdmx.

It is not yet useful. Yet, it provides a catalog of SDMX sources supported by pandaSDMX, and for each source it provides a subcatalog that can read the list of dataflows from the source. Eurostat, e.g., provides thousands of dataflows on everything from road accidents to dairy production and trade balances in European countries.

The SDMX standard comprises very rich metadata on the structure of datasets (dimension names, types, code-lists etc.). I would like to see intake populate a GUI with widgets to select codes for some dimensions and then download a dataset and visualize it. 

I am stuck while thinking about generating user params from that metadata. Where to store it? What to do with it in a script etc.

I could implement _get_schema which would return a schema object with metadata extracted from the SDMX message. But what next? And what would be a sensible workflow from the command line?
  
If you are interested in helping a little, this would be much appreciated. And I guess the large community of pandaSDMX users would applaud intake's features.

Cheers,

Leo

**Comment by martindurant** [comment]
Thanks for getting in touch, I am glad to hear that you are working on this. I have yet to look through your code, but here are a couple of general points. There is enough here that maybe it is worth a one-to-one meeting, particularly since I don't know the details of the pandas implementation.

Intake has two levels of abstraction:
- catalogs, which are lists of named data sources; these sources can themselves be catalogs, so you can use this abstraction to build up arbitrary hierarchical trees. The sources are not instantiated until accessed, so you can preserve laziness. A catalog is dict-like, and cannot be "read" or "partitioned", although its internal entries object might be lazy or support pagination.
- data sources. A catalog initially holds an "entry", and this is instantiated into a source upon access. The user can change the input arguments to the source to produce a new version of the source, and the user-parameters provide a way to indicate to them what is appropriate. User parameters do show up in Intake's rudimentary GUI, so in your case I imagine select-from-a-list widgets for a list of possible options. 

So, it is the responsibility of the catalog to create the user-parameters associated with each of its entries. If these options are easily queried at runtime, that would be fine, especially if you structure the catalogs hierarchically, not to try to fetch too many at once. Or, of course, you might fetch this information and store in the package itself in any format suitable (intake likes YAML, but only because its easy to edit).

**Comment by dr-leo** [comment]
Thanks. This explanation is helpful, even if I have read through the intake docs quite a few times. So I realize that ideally each catalog entry is populated
at instantiation time with the args and user-parameters which will populate the GUI and the yaml file if generated on demand.. Now, the catalog of dataflows
for a given data source can be very long. Eurostat, eg, provides more than 6000 dataflows, each of which references a datastructure definition which references
a bunch of concepts, codelists etc. Populating all catalog entries at catalog creation time with user-parameters generated from these structural metadata
would be overkill.

Btw, the catalog of dataflows is in fact a sub-catalog of the data sources (ECB, Eurostat, UNICEF etc.). But this is accidental detail.

So is there a way to add laziness to each entry in the following sense:

1. At catalog build time, the entries are created from a bare minimum of metadata such as name and description.

2. When the user selects some entry for further inspection, the full set of metadata is downloaded (ie dimension names, types, codelists etc.)

3. Once the user has studied the metadata and set the args according to her interests, she may trigger instantiation of the data source which, through
its read method, downloads the actual data. 

So what I seem to need would be a discover or read method for an entry, right?

Alternatively, I could further split a single SDMX dataflow into entries of a subcatalog whose entries represent sub-cubes. But I dislike this idea, frankly.

To get a flavor of how SDMX and pandasdmx work, see 

https://pandasdmx.readthedocs.io/en/v1.0/example.html?


So a solution, if ugly, could be the following:

Start with a catalog of 6000 dataflows where each entry holds just the minimum of metadata as I wrote above. Each entry instantiates a subcatalog containing
exactly one entry, representing the very same dataflow, but this time populated with the full metadata as user-parameters etc. From there, you can put
together the request for the actual data. Would this make sense?

Another open question is this: Does the GUI allow to select multiple list items (eg. to get a dataset with exchange rates on 5 currency pairs)?

Btw, I failed to get the GUI up and running on Windows. On a separate note, it might be worth noting at this point that due to my blindness I am not a
big fan of GUIs myself.

Happy to schedule a call as you see fit.

Leo

**Comment by martindurant** [comment]
This is a lot of information for me...
Are there really 6000 "flows"? How does one get a list of these or know which one they should access?

Some intake catalog implementations have "lazy" entries, i.e., instead of a dictionary of entries (name-> loading parameters), there is a dict-like which only looks up an entry on access. The understanding is, that the primary access pattern is via direct item access instead of listing or iterating over the whole catalog. This would allow for searching too (but I'm not sure how you get your dataflow names, so I don't know how practical that is).

As a toy example
```python
cat = intake.cat.sdmx # assuming this is "installed", or
cat = intake_sdmx.SDMXCat()
subcat = cat["ECB"]  # this fetches metadata for ecb; could add parameters
list(subcat)  # all of ecb's datasets
subcat.BKN.describe()  # tells of  parameters, as would appear in GUI
s = subcat.BKN(args)  # instantiate data source, with its own metadata
df = s.read()  # to pandas

# the following is possible, but slow
list(cat)
list(cat.items())

# the following API TBD
subcat = cat.search(name or metadata arguments to select flows)
list(subcat)  # concrete set of *flows* that matched the query
```

You see `.describe()` as the method for getting information about an *entry*, although the data source has been instantiated at this time. A datasource with arguments other than default is created from an existing source instance. Instantiating sources should be cheap or free, and there should be a constant cost to `.discover()`to find the schema and metadata; this is called either by the user, or as a necessary step when reading the actual data. 

Alternatively, the top-level catalog could contain entries for all the flows. Instances of the sub-catalogs only get created at access time anyway, so the lookup can happen then - or the sub-catalogs can themselves have lazy entries.  The choice between this and the model above - or a combination of both - depends on how easy the various stages of listing and metadata fetching are. You are obviously the expert on this!

> Does the GUI allow to select multiple list items 

No, the GUI allows for text, integer and choice (one from a number of options) widgets. The widget library `panel` does have an element that would allow it. I'm not sure how I could fit so many options into the available space.

> I am not a big fan of GUIs myself

As primarily a library coder, I agree! However, GUIs and visualization seems to be a good way to attract attention. Honestly, I don't think Intake's GUI is used that often and future specialised versions and integrations with the likes of jupyter-lab are probably more important than anything I write directly into Intake. https://github.com/Xi-CAM/Xi-cam.gui is one specialist Intake datasource browser, for synchrotron experiment data.

**Comment by dr-leo** [comment]
While your above comment has given me a lot of food for thought, I would like to come back to your original proposal for a one-to-one meeting as being the most efficient way to move things forward. If you agree, feel free to suggest a suitable channel, eg. by e-mail at fhaxbox66@gmail.com, where we could agree on a convenient date and time.

**Comment by dr-leo** [comment]
I've implemented much of what we've discussed: 

* use a lazy dict for the dataflows catalog
* populate UserParam instances with allowed codes for each dimension
* instantiate a Source for the datasets with default values for the UserParams.

The following parts should be fairly straightforward:

* search method for dataflows returning a new catalog
* * download the actual dataset based on default values.
Open questions:

* how to overwrite default values with user-provided kwargs? I see that UserParam objects have captured_args etc. attributes. How are these populated by the catalog? I do not see anything helpful in the signature/docs.
* How to store value-level descriptions in a UserParam object? E.g.: 'JPY': 'Japanese Yen'.

Thanks for your great support.

**Comment by martindurant** [comment]
> 1. how to overwrite default values with user-provided kwargs?

The default pattern is for `catalog.item` or `catalog["item"]` to return a data source. This source instance has default values of the user parameters. However, the user can call `.dscribe()` to find out what these parameters are. The data source instance can then be *called* with new values of those user parameters or any other arguments with which the default instance was made (`__call__` aliases method `.configure_new`).

> 2. How to store value-level descriptions in a UserParam object?

The parameter's `.describe()` method would be the obvious place to surface this information. You could even subclass and override the `.validate` method, such that 'JPY' and 'Japanese Yen' are both valid choices and have the same effect. Note that UserParameters have a type, and will coerce inputs to that type - so you could whatever logic you like in here, if you chose to subclass.

> 3. How to select multiple values from a list?

A UserParameter can have a list type, meaning that it will expect and require a list of items. However, the "allowed=" argument currently refers to one choice out of a list. It would not take much code to allow a list type UserParameter in conjunction with an "allowed" list of options. There is no GUI element for this at the moment.

**Comment by dr-leo** [comment]
OK. The above issues seem to be solved, i.e. multiple selection of codes (eg. 'DE+FR+UK', this is properly validated; I've subclassed UserParameter. And code descriptions are stored in the UserParameter alongside the codes itself.

Now I am stuck with the following:
 When instantiating a LocalCatalogEntry (tbd if this is the best subclass), I pass the list of UserParameters. But the resulting DataSource instance does not contain default values for these. And when I re-instantiate the DataSource with some kwarg, say, FREQ='Q', this is ignored.

I had tried doing something like

args={p.name : p.default for p in user_params_list}

and passing args to the LocalCatalogEntry instance. This seemed right as the default values appear in the yaml representation of the dataource. But when I reinstantiate it passing additional args like shown above, these are ignored.

I saw Catalog Entries have some _create_open_args() method. But I don't know when to call it. They should be called when upon instantiation...

**Comment by martindurant** [comment]
Sorry, this week is the Dask Distributed summit, I'm unlikely to have any time to look at this. Please nudge me next week.

**Comment by dr-leo** [comment]
Wonder if you are notified of my thubs-up reaction. So here's the nudge...

**Comment by martindurant** [comment]
> When instantiating a LocalCatalogEntry (tbd if this is the best subclass), I pass the list of UserParameters. But the resulting DataSource instance does not contain default values for these. And when I re-instantiate the DataSource with some kwarg, say, FREQ='Q', this is ignored.

Is this code somewhere? Might be easiest if I have a look.

**Comment by dr-leo** [comment]
Repo:
https://github.com/dr-leo/intake_sdmx

See the only module at top level. Tests are outdated.

The following snippet should be relevant:

```python
# ... list of user-params. loop not included   
params.append(p)
                
        return LocalCatalogEntry(name=flow_id, description=descr, driver=SDMXData, direct_access=True, 
             cache=[], parameters=params, 
             metadata=metadata, catalog_dir='',
             getenv=False, getshell=False, catalog=self)
                 
    
class SDMXData(intake.source.base.DataSource):
    '''
    Driver for SDMX data sets of  a given SDMX dataflow
    '''
    version = __version__
    name = 'sdmx_dataset'
    container = 'dataframe'
    partition_access = True

    def __init__(self, **kwargs):
        super(SDMXData, self).__init__(**kwargs)
        self.name = self.metadata['dataflow_id']
        self.kwargs = kwargs
        
    ```

**Comment by dr-leo** [comment]
What follows is a subjective, high-level summary of the status of intake_sdmx and TODO's for intake and intake_sdmx:

=intake_sdmx: working features=

* catalog of data sources / SDMX services
* subcatalog of dataflows of a selected SDMX service. This works for the main branch, but not for the (allow_multiple_codes feature branch)[https://github.com/dr-leo/intake_sdmx/tree/allow_multiple_code]
* driver for SDMX data sets of a given dataflow. Selection of single code per dimension only.
* read data set and export as pandas DataFrame in various layouts

=TODO's for intake_sdmx=

* test allow_multiple_codes branch and merge. Note: This requires fix of #603.
* update docs 

=TODO's for intake=

* fix #603 and merge. See my comment on that PR
* new release of intake which includes fix for param propagation when deriving DataSource instances using configure_new()

I want intake_sdmx to be relase soon. But this will happen only once a new intake relase is out.

**Comment by martindurant** [comment]
Merge done, release on PyPI (conda-forge will be ready in a couple of hours).

**Comment by dr-leo** [comment]
In v6.4, the mlist type for UserParameter seems to be broken as a list-typed default value is silently converted to str. As a consequence, coercion fails as it expects a list of allowed values. See this [draft PR](https://github.com/dr-leo/intake_sdmx/pull/3) and the following transcript:

```

In [1]: import intake

In [2]: agencies=intake.open_sdmx()

In [3]: agencies
Out[3]: <Intake catalog: None>

In [4]: list(agencies)
Out[4]:
['BBK',
 'BIS',
 'ECB',
 'ESTAT',
 'ILO',
 'INEGI',
 'INSEE',
 'ISTAT',
 'NBB',
 'NB',
 'UNICEF',
 'CD2030',
 'SPC',
 'UNSD',
 'WB',
 'WB_WDI',
 'LSD']

In [5]: ecb=agencies.ECB

In [6]: ecb
Out[6]: <Intake catalog: ECB_SDMX_dataflows>

In [9]: exr=ecb.EXR
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-bd7fa092d223> in <module>
----> 1 exr=ecb.EXR

~\Anaconda3\envs\sdmx377\lib\site-packages\intake\catalog\base.py in __getattr__(self, item)
    348             # Fall back to __getitem__.
    349             try:
--> 350                 return self[item]  # triggers reload_on_change
    351             except KeyError as e:
    352                 raise AttributeError(item) from e

~\Anaconda3\envs\sdmx377\lib\site-packages\intake\catalog\base.py in __getitem__(self, key)
    398             if e.container == 'catalog':
    399                 return e(name=key)
--> 400             return e()
    401         if isinstance(key, str) and '.' in key:
    402             key = key.split('.')

~\Anaconda3\envs\sdmx377\lib\site-packages\intake\catalog\entry.py in __call__(self, persist, **kwargs)
     75             raise ValueError('Persist value (%s) not understood' % persist)
     76         persist = persist or self._pmode
---> 77         s = self.get(**kwargs)
     78         if persist != 'never' and isinstance(s, PersistMixin) and s.has_been_persisted:
     79             from ..container.persist import store

~\Anaconda3\envs\sdmx377\lib\site-packages\intake\catalog\local.py in get(self, **user_parameters)
    287             return self._default_source
    288
--> 289         plugin, open_args = self._create_open_args(user_parameters)
    290         data_source = plugin(**open_args)
    291         data_source.catalog_object = self._catalog

~\Anaconda3\envs\sdmx377\lib\site-packages\intake\catalog\local.py in _create_open_args(self, user_parameters)
    258         open_args = merge_pars(params, user_parameters, self._user_parameters,
 
    
                259                                getshell=self.getshell, getenv=self.getenv,
d--> 260                                client=False)
    261
    262         if len(self._plugin) == 0:

~\Anaconda3\envs\sdmx377\lib\site-packages\intake\catalog\utils.py in merge_pars(params, user_inputs, spec_pars, client, getenv, getshell)
    254         # final validation/coersion
    255         for sp in [p for p in spec_pars if p.name == k]:
--> 256             params[k] = sp.validate(params[k])
    257
    258     return params

~\Anaconda3\envs\sdmx377\lib\site-packages\intake\catalog\local.py in validate(self, value)
    109         """Does value meet parameter requirements?"""
    110         if self.type is not None:
--> 111             value = coerce(self.type, value)
    112
    113         if self.type == "mlist":

~\Anaconda3\envs\sdmx377\lib\site-packages\intake\catalog\utils.py in coerce(dtype, value)
    308             return list(value)
    309         if not isinstance(value, list):
--> 310             raise TypeError("Will not coerce value %s to list", value)
    311         return value
    312     op = COERCION_RULES[dtype]

TypeError: ('Will not coerce value %s to list', "['n/a']")
```

**Comment by martindurant** [comment]
Doing the following:
```
agencies = intake_sdmx.SDMXSources()
ecb=agencies.ECB
exr=ecb.EXR
```
I get a different error:
```
XMLParseError: ConfigError: field "key_member" not yet prepared so type is still a ForwardRef, you might need to call CubeRegion.update_forward_refs().
```
so it's tricky for me to know what's going on.

I have intake from local state of Intake master `0.6.4+34.g7cbd7d1` and intake_sdmx on your allow_multiple_codes branch at f1f7a3a .

**Comment by dr-leo** [comment]
Alas, I cannot reproduce your error. Interestingly, on your side, intake_sdmx seems to request the EXR dataflow from the ECB, whereas on my side, the type error occurs when instantiating the first UserParameter. Hence, I presume that the more recent intake dev version does not yield that error. 

I can only speculate on what might have caused your error. What pandasdmx version are you using? You could do:
```
import pandasdmx
pandasdmx.__version__
```

It should be 1.6.0.

What pydantic version are you using?

It should not be higher than 1.7.2.

Try something like:

```
import pydantic
pydantic.version.version_info()
```

**Comment by martindurant** [comment]
I had thought to check pandasdmx version but not pydantic.  have 1.7.4, so downgrading. Now I see the same exception, so I'll fix it soon.

**Comment by martindurant** [comment]
The following diff does get past this error, but I'm not sure if I like it
```diff
--- a/intake/catalog/utils.py
+++ b/intake/catalog/utils.py
@@ -5,6 +5,7 @@
 # The full license is in the LICENSE file, distributed with this software.
 #-----------------------------------------------------------------------------

+import ast
 from collections.abc import Iterable
 import functools
 import itertools
@@ -313,7 +314,8 @@ def coerce(dtype, value):
         if isinstance(value, (tuple, set, dict)):
             return list(value)
         if not isinstance(value, list):
-            raise TypeError("Will not coerce value %s to list", value)
+            value = ast.literal_eval(value)
+            return list(value)
         return value
     op = COERCION_RULES[dtype]
```

With this change, a source is successfully created, but I see the following error when trying to read:
```
HTTPError: 404 Client Error:  for url: https://sdw-wsrest.ecb.europa.eu/service/data/EXR/n/a.n/a.n/a.n/a.n/a?startPeriod=2021
```
I don't know what the correct URL would look like.

I also happen to notice that the YAML representation of the source contains some pandas timestamp objects, which should probably be datetime-strings .

**Comment by dr-leo** [comment]
I've worked a bit on the intake_sdmx code and was hopeful to release within the next few days. See dr-leo/intake_sdmx#5.

However, it appears that for mlist type user_params, the allowed values do not show up in the yaml:

```


In [1]: from intake_sdmx import *

In [2]: yml = SDMXSources().ECB.EXR.yaml()

In [3]: print(yml)
sources:
  EXR:
    args:
      CURRENCY:
      - '*'
      CURRENCY_DENOM:
      - '*'
      EXR_SUFFIX:
      - '*'
      EXR_TYPE:
      - '*'
      FREQ:
      - '*'
      attributes: ''
      dtype: ''
      endPeriod: !!python/object/apply:pandas._libs.tslibs.timestamps.Timestamp
      - 1609459200000000000
      - null
      - null
      freq_dim: FREQ
      index_type: object
      startPeriod: !!python/object/apply:pandas._libs.tslibs.timestamps.Timestamp
      - 1577836800000000000
      - null
      - null
      storage_options: {}
      time_dim: TIME_PERIOD
    description: Exchange Rates
    driver: intake_sdmx.SDMXData
    metadata:
      catalog_dir: ''
      dataflow_id: EXR
      source_id: ECB
      structure_id: ECB_EXR1

```


Am I missing something?

I realize there are still pandas timestamps. Have to look into this

**Comment by martindurant** [comment]
What you see here is a representation of the instantiated source object, with fixed parameters. You can get the entry it was generated from with either of the following lines:
```
SDMXSources().ECB.EXR.entry
SDMXSources().ECB._entries["EXR"]
```
This entry also has a dict version with `._yaml()` or a string version with `str()`, that you can turn into a standalone source spec in a catalog file, although it has some extra fields you don't need ("name", "container", "plugin", "direct_access")

**Comment by dr-leo** [comment]
OK. So this is a feature. And it kind of makes sense as the yaml of such instance only needs the actual args passed to the instance, not what else would have been allowed.

I'd have to use pyyaml to make a yaml file from exr.entry._yaml(). Right? 

 I guess my question is this: In pandasdmx, the standard way of diving into the codes to select a subcube of the dataflow / dataset is to export it to a DataFrame and use PD methods or some GUI to filter etc. 

What would be the equivalent in intake? Is there a standard way of displaying the allowed values of an entry? Should I add a convenience method to produce a yaml of its entry? Any elegant idea?

**Comment by dr-leo** [comment]
To be more precise: What do I have to do to enable GUI-based code selection? - I haven't tried the GUI so far. So does the GUI inspect entry._yaml() which contains all the code lists from UserParam.allowed?

**Comment by martindurant** [comment]
I'm afraid the GUI doesn't have a selector for mlist-type parameters. This should be added. Honestly, though, the GUI doesn't see much use, so it tends to lag the code.

> I'd have to use pyyaml to make a yaml file from exr.entry._yaml(). Right?

Correct

> Is there a standard way of displaying the allowed values of an entry?

According to the docs `source.describe()` or `entry.describe()` is the official way to do it, which is yet another similar view to the method I suggested above. It's not *particularly* nicely formatted in the standard text output.

**Comment by dr-leo** [comment]
Got you.

I guess the last open question is the one you raised regarding timestamp format in the startPeriod and endPeriod params. They are set to something like
"2021", ie. str(datetime.date().year)

And the type is "datetime".

Here's the code:

```

        # params for startPeriod and endPeriod
        year = date.today().year
        params.extend(
            [
                UserParameter(
                    name="startPeriod",
                    description="startPeriod",
                    type="datetime",
                    default=str(year - 2),
                ),
                UserParameter(
                    name="endPeriod",
                    description="endPeriod",
                    type="datetime",
                    default=str(year - 1),
                ),

```

And this is represented in the yaml as something like

```

      attributes: ''
      dtype: ''
      endPeriod: !!python/object/apply:pandas._libs.tslibs.timestamps.Timestamp
      - 1609459200000000000

```

What to do?

**Comment by martindurant** [comment]
The reason is, that the type is set to `datetime` - but actually they are just strings or numbers (because they are used to format a URL as a string, not passed to a function expecting a datetime object). I would set the type to `str`.

**Comment by dr-leo** [comment]
Thanks. - I've released v0.2.0 on pypi. See the (Documentation)[https://intake_sdmx.readthedocs.io]. Yet to compile as of now.

Can you do a conda pkg or shall I? - For me this is merely a nice-to-have.

Finally, how to announce/advertise?

**Comment by dr-leo** [comment]
v0.2.0 had a bug. I was too tired... Fix is in v0.2.1 on pypi.

Fixed doc build. One error was that msgpack is imported by intake but was not installed. Maybe add msgpack to intake's requirements.
